### PR TITLE
Add support for instrumentation

### PR DESCRIPTION
**Pull request by dasch**
Implements #40.

### CODE REVIEW

**Comment by dasch** [comment]
@bquorning I'm not sure if this can be done in a cleaner way. Care to review?

**Comment by dasch** [comment]
I think it may be possible to use a delegate pattern to separate instrumentation from execution – but in some cases it could be tricky. It would require adding more info to some exceptions, and subscribers would have to correlate multiple events, e.g. the `:attempts` key on `send_messages.kafka` wouldn't be directly available, but rather we'd instrument Transmission as well and emit an event for each attempt. I believe ActiveSupport::Subscriber has support for nested events, i.e. you'd do something like

```ruby
def send_messages(event)
  attempts = event.children.select {|e| e.name == "transmission.kafka" }.count

  # ...  
end

attach_to "kafka"
```

File: kafka.gemspec [hunk 1]
```diff
@@ -32,4 +32,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency "dotenv"
   spec.add_development_dependency "docker-api"
   spec.add_development_dependency "rspec-benchmark"
+  spec.add_development_dependency "activesupport", "~> 4.2.0"
```
Comment (bquorning) [inline kafka.gemspec]: Doesn’t it work with the Rails 5 betas?

**Comment by bquorning** [comment]
:+1: for the current approach.

File: kafka.gemspec (hunk 1 above)
Comment (dasch) [inline kafka.gemspec]: I always use the latest stable release from libs.

**Comment by dasch** [comment]
I'm taking this in a different direction: instead of instrumenting the high-level parts, I'm instrumenting the low-level ones, starting with Connection.

Individual requests are tracked, and the following keys are reported in the payload:

* `:api`: a symbol referencing which API was invoked.
* `:request_size`: the number of bytes sent to the server.
* `:response_size` the number of bytes received from the server.

In case of a connection error, the `:exception` key will hold the name of the exception class as well as any error message.

**Pull request merged**